services/worker-manager/src/api.js

```diff
@@ -11,231 +8,160 @@ let builder = new APIBuilder({
-    res.reportError('RequestConflict', 'worker configuration id rest parameter must match body');
-  }
+  await req.authorize({name, provider: providerName});
 
-  try {
-    await this.datastore.get('worker-configurations', id);
-  } catch (err) {
-    if (err instanceof errors.InvalidDatastoreKey) {
-      return res.reportError('ResourceNotFound', `${id} is unknown`);
-    }
-    throw err;
+  const provider = this.providers[providerName];
+  if (!provider) {
+    return res.reportError('InputError', 'Invalid Provider', {
+      provider: providerName,
+      validProviders: Object.keys(this.providers),
+    });
   }
 
-  let workerTypeConflicts = await workerTypeAlreadyTracked(this.datastore, workerConfiguration);
-  if (workerTypeConflicts.length > 0) {
-    return res.reportError('RequestConflict', 'worker type conflicts', {
-      conflicts: workerTypeConflicts,
-    });
+  const workerType = await this.WorkerType.load({
+    name,
+  }, true);
+  if (!workerType) {
+    return res.reportError('ResourceNotFound', 'WorkerType does not exist', {});
   }
 
-  await this.datastore.set('worker-configurations', id, req.body);
+  await workerType.modify(wt => {
+    wt.config = input.config; // TODO: validate
+    wt.description = input.description;
+    wt.provider = providerName;
+    wt.owner = input.owner;
+    wt.lastModifed = new Date().toJSON();
+  });
 
-  res.reply();
+  res.reply(workerType.serializable());
 });
 
 builder.declare({
   method: 'get',
-  route: '/worker-configurations/:workerConfigurationId',
-  name: 'getWorkerConfiguration',
-  title: 'Get Worker Configuration',
+  route: '/workertype/:name',
+  name: 'workerType',
+  title: 'Get WorkerType',
   stability: APIBuilder.stability.experimental,
-  output: 'worker-configuration.yml',
+  output: 'workertype-full.yml',
   description: [
-    'Get a worker configuration',
+    'TODO',
   ].join('\n'),
 }, async function(req, res) {
-  let id = req.params.workerConfigurationId;
-  try {
-    res.reply(await this.datastore.get('worker-configurations', id));
-  } catch (err) {
-    if (err instanceof errors.InvalidDatastoreKey) {
-      return res.reportError('ResourceNotFound', `${id} is unknown`);
-    }
-    throw err;
+  const {name} = req.params;
+
+  const workerType = await this.WorkerType.load({
+    name,
+  }, true);
+  if (!workerType) {
+    return res.reportError('ResourceNotFound', 'WorkerType does not exist', {});
   }
+  res.reply(workerType.serializable());
 });
 
 builder.declare({
   method: 'delete',
-  route: '/worker-configurations/:workerConfigurationId',
-  name: 'removeWorkerConfiguration',
-  title: 'Remove Worker Configuration',
+  route: '/workertype/:name',
+  name: 'deleteWorkerType',
+  title: 'Delete WorkerType',
+  scopes: 'worker-manager:delete-worker-type:<name>',
   stability: APIBuilder.stability.experimental,
   description: [
-    'Get a worker configuration',
+    'TODO',
   ].join('\n'),
 }, async function(req, res) {
```

Comment:
What do you think about marking the workerType as deleted and then relying on the provider to delete it when it's ready?  That would give it a chance to delete any leftover resources, terminate instances, etc.

Reply:
That could work too. In the next PR you'll see how I've done it with things as they are now. We can change it in that case if you don't like it.